File: scripts/sync-scheduler.py
# Firebase → PostgreSQL Sync Scheduler
# This script starts the sync scheduler and handles process management

import os
import sys
import time
from datetime import datetime
from src.sync.enhanced_sync_scheduler import EnhancedSyncScheduler

def print_status(message):
    print(f"[INFO] {message}")

def print_warning(message):
    print(f"[WARNING] {message}")

def print_error(message):
    print(f"[ERROR] {message}")

def check_firebase_env():
    required_vars = [
        "VITE_FIREBASE_APIKEY",
        "VITE_FIREBASE_PROJECTID",
        "VITE_FIREBASE_AUTHDOMAIN",
        "VITE_FIREBASE_STORAGEBUCKET",
        "VITE_FIREBASE_MESSAGINGSENDERID",
        "VITE_FIREBASE_APPID"
    ]

    missing_vars = [var for var in required_vars if not os.getenv(var)]

    if missing_vars:
        print_error("Firebase environment variables are not set:")
        for var in missing_vars:
            print_error(f"  {var}")
        return False
    return True

def check_postgres_env():
    required_vars = [
        "VITE_POSTGRES_USER",
        "VITE_POSTGRES_DB"
    ]

    missing_vars = [var for var in required_vars if not os.getenv(var)]

    if missing_vars:
        print_warning("PostgreSQL environment variables are not fully set:")
        for var in missing_vars:
            print_warning(f"  {var}")
        print_warning("Using default values. Ensure these are correct for production.")
    return True

def start_sync_scheduler():
    print_status("Starting Firebase → PostgreSQL sync scheduler...")

    # Check environment variables
    if not check_firebase_env():
        sys.exit(1)

    check_postgres_env()

    # Navigate to the project root
    os.chdir(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

    # Start the sync scheduler
    print_status("Starting sync scheduler...")

    try:
        scheduler = EnhancedSyncScheduler()
        scheduler.start()
    except KeyboardInterrupt:
        print_status("Sync scheduler stopped by user.")
        sys.exit(0)
    except Exception as e:
        print_error(f"Failed to start sync scheduler: {e}")
        sys.exit(1)

def show_sync_status():
    print_status("Checking sync scheduler status...")

    try:
        from src.sync.enhanced_sync_scheduler import EnhancedSyncScheduler
        scheduler = EnhancedSyncScheduler()
        print(scheduler.getSyncStatus())
    except Exception as e:
        print_error(f"Failed to get sync status: {e}")

def trigger_manual_sync(user_id):
    if not user_id:
        print_error("User ID is required for manual sync.")
        print_error("Usage: python sync-scheduler.py manual-sync <user-id>")
        sys.exit(1)

    print_status(f"Triggering manual sync for user: {user_id}")

    try:
        from src.sync.enhanced_sync_scheduler import EnhancedSyncScheduler
        scheduler = EnhancedSyncScheduler()
        result = scheduler.triggerSync(user_id)
        print(f"Manual sync result: {result}")
    except Exception as e:
        print_error(f"Manual sync error: {e}")

def show_help():
    print("Firebase → PostgreSQL Sync Scheduler")
    print()
    print("Usage: python sync-scheduler.py [COMMAND]")
    print()
    print("Commands:")
    print("  start          Start the sync scheduler")
    print("  status         Show sync scheduler status")
    print("  manual-sync    Trigger manual sync for a specific user")
    print("  help           Show this help message")
    print()
    print("Examples:")
    print("  python sync-scheduler.py start")
    print("  python sync-scheduler.py manual-sync user_12345")

if __name__ == "__main__":
    command = sys.argv[1] if len(sys.argv) > 1 else "start"

    if command == "start":
        start_sync_scheduler()
    elif command == "status":
        show_sync_status()
    elif command == "manual-sync":
        user_id = sys.argv[2] if len(sys.argv) > 2 else None
        trigger_manual_sync(user_id)
    elif command in ("help", "--help", "-h"):
        show_help()
    else:
        print_error(f"Unknown command: {command}")
        show_help()
        sys.exit(1)

File: scripts/sync-scheduler.js
#!/usr/bin/env node

/**
 * Firebase → PostgreSQL Sync Scheduler CLI
 * Node.js version of the sync scheduler command-line interface
 */

import enhancedSyncScheduler from '../src/sync/enhanced-sync-scheduler.js';
import fs from 'fs';
import path from 'path';

// Colors for output
const colors = {
  reset: '\x1b[0m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m'
};

function printStatus(message) {
  console.log(`${colors.green}[INFO]${colors.reset} ${message}`);
}

function printWarning(message) {
  console.log(`${colors.yellow}[WARNING]${colors.reset} ${message}`);
}

function printError(message) {
  console.log(`${colors.red}[ERROR]${colors.reset} ${message}`);
}

function checkFirebaseEnv() {
  const requiredVars = [
    'VITE_FIREBASE_APIKEY',
    'VITE_FIREBASE_PROJECTID',
    'VITE_FIREBASE_AUTHDOMAIN',
    'VITE_FIREBASE_STORAGEBUCKET',
    'VITE_FIREBASE_MESSAGINGSENDERID',
    'VITE_FIREBASE_APPID'
  ];

  const missingVars = requiredVars.filter(varName => !process.env[varName]);

  if (missingVars.length > 0) {
    printError('Firebase environment variables are not set:');
    missingVars.forEach(varName => printError(`  ${varName}`));
    return false;
  }
  return true;
}

function checkPostgresEnv() {
  const requiredVars = [
    'VITE_POSTGRES_USER',
    'VITE_POSTGRES_DB'
  ];

  const missingVars = requiredVars.filter(varName => !process.env[varName]);

  if (missingVars.length > 0) {
    printWarning('PostgreSQL environment variables are not fully set:');
    missingVars.forEach(varName => printWarning(`  ${varName}`));
    printWarning('Using default values. Ensure these are correct for production.');
  }
  return true;
}

async function startSyncScheduler() {
  printStatus('Starting Firebase → PostgreSQL sync scheduler...');

  // Check environment variables
  if (!checkFirebaseEnv()) {
    process.exit(1);
  }
  checkPostgresEnv();

  // Navigate to the project root
  const scriptDir = __dirname;
  const projectRoot = path.resolve(scriptDir, '../..');
  process.chdir(projectRoot);

  // Start the sync scheduler
  printStatus('Starting sync scheduler...');

  try {
    await enhancedSyncScheduler.start();
  } catch (error) {
    printError(`Failed to start sync scheduler: ${error.message}`);
    process.exit(1);
  }
}

async function showSyncStatus() {
  printStatus('Checking sync scheduler status...');

  try {
    const scheduler = new EnhancedSyncScheduler();
    console.log(scheduler.getSyncStatus());
  } catch (error) {
    printError(`Failed to get sync status: ${error.message}`);
  }
}

async function triggerManualSync(userId) {
  if (!userId) {
    printError('User ID is required for manual sync.');
    printError('Usage: node sync-scheduler.js manual-sync <user-id>');
    process.exit(1);
  }

  printStatus(`Triggering manual sync for user: ${userId}`);

  try {
    const scheduler = new EnhancedSyncScheduler();
    const result = await scheduler.triggerSync(userId);
    console.log('Manual sync result:', result);
  } catch (error) {
    printError(`Manual sync error: ${error.message}`);
  }
}

function showHelp() {
  console.log('Firebase → PostgreSQL Sync Scheduler');
  console.log();
  console.log('Usage: node sync-scheduler.js [COMMAND]');
  console.log();
  console.log('Commands:');
  console.log('  start          Start the sync scheduler');
  console.log('  status         Show sync scheduler status');
  console.log('  manual-sync    Trigger manual sync for a specific user');
  console.log('  help           Show this help message');
  console.log();
  console.log('Examples:');
  console.log('  node sync-scheduler.js start');
  console.log('  node sync-scheduler.js manual-sync user_12345');
}

// Main script logic
const command = process.argv[2] || 'start';

switch (command) {
  case 'start':
    startSyncScheduler();
    break;
  case 'status':
    showSyncStatus();
    break;
  case 'manual-sync':
    const userId = process.argv[3];
    triggerManualSync(userId);
    break;
  case 'help':
  case '--help':
  case '-h':
    showHelp();
    break;
  default:
    printError(`Unknown command: ${command}`);
    showHelp();
    process.exit(1);
}

// Make the script executable
if (require.main === module) {
  // This is the main module, do nothing special
}